# LAB 1 - INTRO TO PYTHON

This lab is comprised of three parts:

- 1) Introduction to Jupyter Notebooks

- 2) Python Language and NumPy Library

- 3) Data Manipulation with Pandas Library


## 1) INTRODUCTION TO JUPYTER NOTEBOOKS

Open the file "Lab1.ipynb"

`"File -> Open..."` 

Navigate to where you saved the files you downloaded for this class and double click on the "Lab1.ipynb" file

### Running notebook cells

The notebook is divided into cells. Each cell can contain texts, codes or html scripts. Running a non-code cell simply advances to the next cell. Make sure to type commands always in a Code cell. You can verify this by checking the scroll-down window in the above toolbar menu.

To run a code cell use `shift + enter` or `ctrl + enter`. 

Try the following commends:

1) shift + enter run cell, select below

2) ctrl + enter run cell

3) option + enter run cell, insert below

In [13]:
8*6

48

In [14]:
2**16

65536

Incomplete command, Jupyter will display a `SyntaxError`

In [15]:
2^

SyntaxError: invalid syntax (<ipython-input-15-0deb74d06d3d>, line 1)

Exercise 1.1 Your turn:

In [16]:
# EXERCISE: Compute 284455 divided by 3.67778
284455/3.67778

77344.21308506763

Note: In Python, any text following a hash sign in a code cell is a comment

### Interrupting the kernel

For debugging, often we would like to interupt the current running process. This can be done by pressing the stop button. 

When a processing is running, the circle on the right upper corner is filled. When idle, the circle is empty.

Interrupting sometimes does not work. You can reset the state by restarting the kernel. This is done by clicking Kernel/Restart or the Refresh button in the toolbar above.

### Undoing

To undo changes in each cell, hit `Command-z` for Mac and `Ctrl-z` for Windows.
To undo `Delete Cell`, select `Edit->Undo Delete Cell`.

### Saving the notebook

To save your notebook, either select `"File->Save and Checkpoint"` or hit `Command-s` for Mac and `Ctrl-s` for Windows

### Other Notebook tips
- To add a new cell, either select `"Insert->Insert New Cell Below"` or click the white plus button
- You can change the cell mode from code to text in the pulldown menu by selecting `Markdown`.
- `Help->Keyboard Shortcuts` has a list of keyboard shortcuts

In [17]:
3*9

27

## 2) PYTHON LANGUAGE AND NUMPY LIBRARY

##  Data Types

### Floats and Integers

In [18]:
x = 4
print(x, type(x))

4 <class 'int'>


In [19]:
x = 1 / 4
print(x, type(x))

0.25 <class 'float'>


### Strings

Double quotes and single quotes are the same thing. Both represent strings. `'+'` concatenates strings

In [20]:
"Jan " + '2023'

'Jan 2023'

### Lists

A list is a mutable collection of data, which means that we can change it after it is created. A list can be created using square brackets []


Important functions: 
- `'+'` appends lists. 
- `len(x)` returns the length of a list.

In [21]:
x = ["Jan"] + [2, 0, 2, 3]
print(x)

['Jan', 2, 0, 2, 3]


In [22]:
print(len(x))

5


In [23]:
x[5]

IndexError: list index out of range

In [24]:
x[4]=2

In [25]:
print(x)

['Jan', 2, 0, 2, 2]


### Tuples

A tuple is an immutable collection of data. They can be created using round brackets (). 
They are usually used as inputs and outputs to functions.

In [26]:
t = ("J", "A", "N") + (2, 0, 2, 3)
print(t)

('J', 'A', 'N', 2, 0, 2, 3)


In [27]:
# cannot do assignment to a tuple after creation - it's immutable
t[4] = 3 # will cause error

# Note: errors in notebook appear inline

TypeError: 'tuple' object does not support item assignment

## Functions and Variables

A function can take in several arguments or inputs, and returns an output value.
Python has some built-in functions:

In [28]:
abs(-65)

65

In [29]:
max([2, 4, 2])

4

In [30]:
# Get help on any function:
max?

Docstring:
max(iterable, *[, default=obj, key=func]) -> value
max(arg1, arg2, *args, *[, key=func]) -> value

With a single iterable argument, return its biggest item. The
default keyword-only argument specifies an object to return if
the provided iterable is empty.
With two or more arguments, return the largest argument.
Type:      builtin_function_or_method


Basic variable naming rules: 
- Don't use spaces (underscores or capital letters instead)
- Don't start names with a number
- Variable names are case sensitive - capital and lowercase letters are different

Exercise 1.2

In [31]:
# EXERCISE: 
# Create a variable called "SecondsDay" that is equal to the number of 
# seconds in a day, and output its value.

SecondsDay=24*60*60
print(SecondsDay)

86400


### User-defined Functions
We can define functions ourselves, by using def and passing the expected inputs, as well as stating the returned output. In this example we create a function that takes two numbers x and y, and returns the sum.

In [32]:
def my_function(x, y):
    
    result = x+y
    
    return result

In [33]:
my_function(5, 3)

8

## Linear Algebra with Numpy

The numpy array, aka an "ndarray", is like a list with multidimensional support and more functions.
https://numpy.org/doc/stable/reference/routines.linalg.html

Important NumPy Array functions:

- `.shape` returns the dimensions of the array.

- `.ndim` returns the number of dimensions. 

- `.size` returns the number of entries in the array.

- `len()` returns the first dimension.


To use functions in NumPy, we have to import NumPy to our workspace. This is done by the command `import numpy`. By convention, we rename `numpy` as `np` for convenience.

### Arrays

In [34]:
import numpy as np

a = np.array([1,2,3])
a.shape

(3,)

In [35]:
# Multiplication of a constant and a vector
2*a

array([2, 4, 6])

In [36]:
# Element-wise multiplication
b = np.array([3,3,3])
np.multiply(a,b)

array([3, 6, 9])

Note: The two examples we did above touch the essense of two important concepts in NumPy array -- vectorization and broadcasting.

It is a good habit to employ vectorization and broadcasting whenever possible when dealing with linear algebra in NumPy. It will avoid unnecessary loops and significantly improve the efficiency of your code. 

Read more of the documentation at https://numpy.org/doc/stable/user/basics.broadcasting.html

In [37]:
# Inner product
inner_product = np.dot(a,b)
print(inner_product)

18


### Slicing

Numpy uses pass-by-reference semantics so it creates views into the existing array, without implicit copying. This is particularly helpful with very large arrays because copying can be slow.

In [38]:
x = ['Jan', 2, 0, 2, 2]
y = x[0:4]
print(y)
y[1]=3
print(y)
print(x)

['Jan', 2, 0, 2]
['Jan', 3, 0, 2]
['Jan', 2, 0, 2, 2]


In [39]:
y = x[0:4]
print(y)

['Jan', 2, 0, 2]


In [40]:
x

['Jan', 2, 0, 2, 2]

In [41]:
y[1]=3

In [42]:
y

['Jan', 3, 0, 2]

In [43]:
x

['Jan', 2, 0, 2, 2]

Notes: since slicing does not copy the array, any change made to `y` would also change `x`. To create an object `y` that does not bind with the original object `x`, one need to make a copy of `x`. 

To achieve this, one should use `.copy()` from the `copy` library. (Documentation: https://docs.python.org/3/library/copy.html)

### Matrices

In [44]:
# Create a matrix
A = np.array([[1, 2, 8],
             [3, 2, 9]])
print(A)

[[1 2 8]
 [3 2 9]]


In [45]:
# Matrix multiplication
B = np.array([[1, 2],
              [3, 8],
              [2, 9]])

# There are two ways to perform matrix multiplication:
print(np.matmul(A,B))

# Alternatively:
print(A@B)

[[ 23  90]
 [ 27 103]]
[[ 23  90]
 [ 27 103]]


In [46]:
# Transpose a matrix
A.T

array([[1, 3],
       [2, 2],
       [8, 9]])

In [47]:
# Compute the inverse
C = np.array([[1, 2],
             [3, 2]])
D = np.linalg.inv(C)
C@D # note: the off-diagonal entries are essentially zero, so the output is an identity matrix.

array([[1.00000000e+00, 1.11022302e-16],
       [2.22044605e-16, 1.00000000e+00]])

In [48]:
# Reshape 1-d NumPy Array to 2-d matrix
X = np.array([1,2,3])
print(X)
print(X.shape)

Y = np.reshape(X,(-1,1))
print(Y)
print(Y.shape)
# This technique is useful when you want to convert a 1-d vector to a 2-d array (a matrix with only 1 column).  

[1 2 3]
(3,)
[[1]
 [2]
 [3]]
(3, 1)


# 3) DATA MANIPULATION WITH PANDAS LIBRARY

`pandas` is designed to make it easier to work with structured data. Most of the analyses you might perform will likely involve using tabular data, e.g., from .csv files or relational databases (e.g., SQL). The `DataFrame` object in `pandas` is "a two-dimensional tabular, column-oriented data structure with both row and column labels."

If you're curious:

>The `pandas` name itself is derived from *panel data*, an econometrics term for multidimensional structured data sets, and *Python data analysis* itself. After getting introduced, you can consult the full [`pandas` documentation](http://pandas.pydata.org/pandas-docs/stable/).

### Setting the working directory

Before loading the data, let's begin by setting the right working directory. In order to change the working directory, we use the `os` library

In [49]:
import os
os.getcwd()

'/Users/ghitaalami/235'

Use the following line to change the working directory to the path where your python files and data files are saved in

In [50]:
os.chdir("")

FileNotFoundError: [Errno 2] No such file or directory: ''

### Loading CSV files

Now we can use the `pandas` library to load the data. Import `pandas` using the conventional abbreviation and call the `read_csv` method on your file's path name

In [51]:
import pandas as pd

# WHO = pd.read_csv("WHO.csv")
WHO = pd.read_csv("WHO.csv", encoding = "ISO-8859-1") ##check this encoding, also encoding = 'unicode_escape'

### The Dataframe

In [52]:
WHO

Country                 Region  Population  \
0                           Afghanistan  Eastern Mediterranean       29825   
1                               Albania                 Europe        3162   
2                               Algeria                 Africa       38482   
3                               Andorra                 Europe          78   
4                                Angola                 Africa       20821   
..                                  ...                    ...         ...   
189  Venezuela (Bolivarian Republic of)               Americas       29955   
190                            Viet Nam        Western Pacific       90796   
191                               Yemen  Eastern Mediterranean       23852   
192                              Zambia                 Africa       14075   
193                            Zimbabwe                 Africa       13724   

     Under15  Over60  FertilityRate  LifeExpectancy  ChildMortality  \
0      47.42    3.82           5.40              60            98.5   
1      21.33   14.93           1.75              74            16.7   
2      27.42    7.17           2.83              73            20.0   
3      15.20   22.86            NaN              82             3.2   
4      47.58    3.84           6.10              51           163.5   
..       ...     ...            ...             ...             ...   
189    28.84    9.17           2.44              75            15.3   
190    22.87    9.32           1.79              75            23.0   
191    40.72    4.54           4.35              64            60.0   
192    46.73    3.95           5.77              55            88.5   
193    40.24    5.68           3.64              54            89.8   

     CellularSubscribers  LiteracyRate      GNI  PrimarySchoolEnrollmentMale  \
0                  54.26           NaN   1140.0                          NaN   
1                  96.39           NaN   8820.0                          NaN   
2                  98.99           NaN   8310.0                         98.2   
3                  75.49           NaN      NaN                         78.4   
4                  48.38          70.1   5230.0                         93.1   
..                   ...           ...      ...                          ...   
189                97.78           NaN  12430.0                         94.7   
190               143.39          93.2   3250.0                          NaN   
191                47.05          63.9   2170.0                         85.5   
192                60.59          71.2   1490.0                         91.4   
193                72.13          92.2      NaN                          NaN   

     PrimarySchoolEnrollmentFemale  
0                              NaN  
1                              NaN  
2                             96.4  
3                             79.4  
4                             78.2  
..                             ...  
189                           95.1  
190                            NaN  
191                           70.5  
192                           93.9  
193                            NaN  

[194 rows x 13 columns]

In [53]:
# Structure of the data
WHO.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 194 entries, 0 to 193
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country                        194 non-null    object 
 1   Region                         194 non-null    object 
 2   Population                     194 non-null    int64  
 3   Under15                        194 non-null    float64
 4   Over60                         194 non-null    float64
 5   FertilityRate                  183 non-null    float64
 6   LifeExpectancy                 194 non-null    int64  
 7   ChildMortality                 194 non-null    float64
 8   CellularSubscribers            184 non-null    float64
 9   LiteracyRate                   103 non-null    float64
 10  GNI                            162 non-null    float64
 11  PrimarySchoolEnrollmentMale    101 non-null    float64
 12  PrimarySchoolEnrollmentFemale  101 non-null    flo

In [54]:
# Recent statistics from the World Health Organization (WHO)
# The variables are: 
# the name of the country
# the region the country is in
# the population in thousandsa
# the percentage of the population under 15 and over 60
# the fertility rate (average number of children per woman)
# the Life Expectancy in years
# the Child Mortality rate (the number of children who die by age 5 per 1000 births)
# the number of cellular subscribers per 100 population
# the literacy rate among adults aged >= 15
# the gross national income per capita
# the percentage of male children enrolled in primary school
# the percentage of female children enrolled in primary school

In [55]:
# Statistical summary of the data:
WHO.describe()

Population     Under15      Over60  FertilityRate  LifeExpectancy  \
count  1.940000e+02  194.000000  194.000000     183.000000      194.000000   
mean   3.635997e+04   28.732423   11.163660       2.940656       70.010309   
std    1.379031e+05   10.534573    7.149331       1.480984        9.259075   
min    1.000000e+00   13.120000    0.810000       1.260000       47.000000   
25%    1.695750e+03   18.717500    5.200000       1.835000       64.000000   
50%    7.790000e+03   28.650000    8.530000       2.400000       72.500000   
75%    2.453525e+04   37.752500   16.687500       3.905000       76.000000   
max    1.390000e+06   49.990000   31.920000       7.580000       83.000000   

       ChildMortality  CellularSubscribers  LiteracyRate           GNI  \
count      194.000000           184.000000    103.000000    162.000000   
mean        36.148969            93.641522     83.710680  13320.925926   
std         37.992935            41.400447     17.530645  15192.988650   
min          2.200000             2.570000     31.100000    340.000000   
25%          8.425000            63.567500     71.600000   2335.000000   
50%         18.600000            97.745000     91.800000   7870.000000   
75%         55.975000           120.805000     97.850000  17557.500000   
max        181.600000           196.410000     99.800000  86440.000000   

       PrimarySchoolEnrollmentMale  PrimarySchoolEnrollmentFemale  
count                   101.000000                     101.000000  
mean                     90.850495                      89.632673  
std                      11.017147                      12.817614  
min                      37.200000                      32.500000  
25%                      87.700000                      87.300000  
50%                      94.700000                      95.100000  
75%                      98.100000                      97.900000  
max                     100.000000                     100.000000

In [56]:
# Display a few data points at the "head" (start) of the dataset, i.e. the first few records
# By default, the first 5 records are shown; this can be overwritten by specificing the number of records in the paranthesis.
WHO.head(2) # try WHO.head(2)

Country                 Region  Population  Under15  Over60  \
0  Afghanistan  Eastern Mediterranean       29825    47.42    3.82   
1      Albania                 Europe        3162    21.33   14.93   

   FertilityRate  LifeExpectancy  ChildMortality  CellularSubscribers  \
0           5.40              60            98.5                54.26   
1           1.75              74            16.7                96.39   

   LiteracyRate     GNI  PrimarySchoolEnrollmentMale  \
0           NaN  1140.0                          NaN   
1           NaN  8820.0                          NaN   

   PrimarySchoolEnrollmentFemale  
0                            NaN  
1                            NaN

In [57]:
# Display the last few records.
WHO.tail()

Country                 Region  Population  \
189  Venezuela (Bolivarian Republic of)               Americas       29955   
190                            Viet Nam        Western Pacific       90796   
191                               Yemen  Eastern Mediterranean       23852   
192                              Zambia                 Africa       14075   
193                            Zimbabwe                 Africa       13724   

     Under15  Over60  FertilityRate  LifeExpectancy  ChildMortality  \
189    28.84    9.17           2.44              75            15.3   
190    22.87    9.32           1.79              75            23.0   
191    40.72    4.54           4.35              64            60.0   
192    46.73    3.95           5.77              55            88.5   
193    40.24    5.68           3.64              54            89.8   

     CellularSubscribers  LiteracyRate      GNI  PrimarySchoolEnrollmentMale  \
189                97.78           NaN  12430.0                         94.7   
190               143.39          93.2   3250.0                          NaN   
191                47.05          63.9   2170.0                         85.5   
192                60.59          71.2   1490.0                         91.4   
193                72.13          92.2      NaN                          NaN   

     PrimarySchoolEnrollmentFemale  
189                           95.1  
190                            NaN  
191                           70.5  
192                           93.9  
193                            NaN

### Subsets of data

In [58]:
# find the subset with only the countries in Europe
WHO_Europe = WHO[WHO['Region'] == 'Europe']
WHO_Europe.head()

Country  Region  Population  Under15  Over60  FertilityRate  \
1      Albania  Europe        3162    21.33   14.93           1.75   
3      Andorra  Europe          78    15.20   22.86            NaN   
7      Armenia  Europe        2969    20.34   14.06           1.74   
9      Austria  Europe        8464    14.51   23.52           1.44   
10  Azerbaijan  Europe        9309    22.25    8.24           1.96   

    LifeExpectancy  ChildMortality  CellularSubscribers  LiteracyRate  \
1               74            16.7                96.39           NaN   
3               82             3.2                75.49           NaN   
7               71            16.4               103.57          99.6   
9               81             4.0               154.78           NaN   
10              71            35.2               108.75           NaN   

        GNI  PrimarySchoolEnrollmentMale  PrimarySchoolEnrollmentFemale  
1    8820.0                          NaN                            NaN  
3       NaN                         78.4                           79.4  
7    6100.0                          NaN                            NaN  
9   42050.0                          NaN                            NaN  
10   8960.0                         85.3                           84.1

In [59]:
WHO_Europe.count()

Country                          53
Region                           53
Population                       53
Under15                          53
Over60                           53
FertilityRate                    50
LifeExpectancy                   53
ChildMortality                   53
CellularSubscribers              51
LiteracyRate                     26
GNI                              48
PrimarySchoolEnrollmentMale      38
PrimarySchoolEnrollmentFemale    38
dtype: int64

In [60]:
# Use compound boolean operators in the conditional expression
# Use & for and; | for or. 
WHO_AsiaEurope = WHO[(WHO['Region'] == 'Europe') | (WHO['Region'] == 'South-East Asia') | (WHO['Region'] == "Eastern Mediterranean")] 
WHO_AsiaEurope.head()

Country                 Region  Population  Under15  Over60  \
0  Afghanistan  Eastern Mediterranean       29825    47.42    3.82   
1      Albania                 Europe        3162    21.33   14.93   
3      Andorra                 Europe          78    15.20   22.86   
7      Armenia                 Europe        2969    20.34   14.06   
9      Austria                 Europe        8464    14.51   23.52   

   FertilityRate  LifeExpectancy  ChildMortality  CellularSubscribers  \
0           5.40              60            98.5                54.26   
1           1.75              74            16.7                96.39   
3            NaN              82             3.2                75.49   
7           1.74              71            16.4               103.57   
9           1.44              81             4.0               154.78   

   LiteracyRate      GNI  PrimarySchoolEnrollmentMale  \
0           NaN   1140.0                          NaN   
1           NaN   8820.0                          NaN   
3           NaN      NaN                         78.4   
7          99.6   6100.0                          NaN   
9           NaN  42050.0                          NaN   

   PrimarySchoolEnrollmentFemale  
0                            NaN  
1                            NaN  
3                           79.4  
7                            NaN  
9                            NaN

### Saving dataframe to CSV file

In [61]:
WHO_AsiaEurope.to_csv("WHO_AsiaEurope.csv")

Exercise 1.3

In [64]:
# Your turn:
# How many countries have population greater than 50 million? 
who_pop=WHO[WHO["Population"]>50000]
who_pop.count()[0]

25

### More Data Analysis

To access a variable in a data frame, you always have to link it to the data frame and call it using square brackets and pass it's name as a string.

In [63]:
# This will give you an error!
LifeExpectancy

NameError: name 'LifeExpectancy' is not defined

In [65]:
# Now, run this.
WHO['LifeExpectancy']

0      60
1      74
2      73
3      82
4      51
       ..
189    75
190    75
191    64
192    55
193    54
Name: LifeExpectancy, Length: 194, dtype: int64

### Statistics

In [66]:
# Statistics of a variable
print((WHO['LifeExpectancy'].mean()))
print((WHO['LifeExpectancy'].max()))
print((WHO['LifeExpectancy'].min()))

70.01030927835052
83
47


In [67]:
# Standard deviation
WHO['LifeExpectancy'].std
WHO['LifeExpectancy'].describe()

count    194.000000
mean      70.010309
std        9.259075
min       47.000000
25%       64.000000
50%       72.500000
75%       76.000000
max       83.000000
Name: LifeExpectancy, dtype: float64

In [68]:
WHO['GNI'].describe()
# what's different here?

count      162.000000
mean     13320.925926
std      15192.988650
min        340.000000
25%       2335.000000
50%       7870.000000
75%      17557.500000
max      86440.000000
Name: GNI, dtype: float64

Here there are less instances (162) compared to the previous cell. This can be due to the presence of NaN values.

In [75]:
# Identify countries corresponding to max and min
idx_min = WHO['LifeExpectancy'].argmin()
print(WHO.iloc[idx_min]['Country'])

idx_max = WHO['LifeExpectancy'].argmax()
print(WHO['Country'][idx_max])

Sierra Leone
Cªte d'Ivoire


Exercise 1.4

In [ ]:
# Exercise:
# What is the largest population value among all countries?
# Which country has the largest population?

idx_max_pop=WHO['Population'].argmax()
print(WHO['Population'][idx_max_pop])
print(WHO['Country'][idx_max_pop])


1390000
China


### Dealing with missing data

In [ ]:
# Dealing with NA
# Try:
WHO['LiteracyRate'].head()

0     NaN
1     NaN
2     NaN
3     NaN
4    70.1
Name: LiteracyRate, dtype: float64

In [ ]:
WHO.dropna(subset=['LiteracyRate'], inplace=True)
WHO['LiteracyRate'].head()
# Note: setting "inplace = True" will modify the original dataframe. 
# Alternatively, setting "inplace = False" will generate a new dataframe.

4     70.1
5     99.0
6     97.8
7     99.6
12    91.9
Name: LiteracyRate, dtype: float64

# References
- [1] Special thanks to the [EECS127 Fall 2019](https://inst.eecs.berkeley.edu/~ee127/fa19/) for providing a great starting point for Intro to Jupyter
- [2] D-lab intro to pandas
- [3] The official Python 3 language documentation. [Link](https://docs.python.org/3/).
- [4] The official numpy and scipy documentation. [Link](https://docs.scipy.org/doc/).


